### ISSUE TITLE
S3 create event notification not propagating trace context to SQS destination

### ISSUE THREAD

**Issue by afayes** [aws/aws-xray-sdk-node#445]
I am working on an application where we want to integrate X-Ray traces.

The flow of services looks like:
API Gateway -> Lambda -> S3 -> SQS -> Lambda -> API Gateway

X-Ray tracing from API Gateway to Lambda works fine.

to integrate S3 I instrumented the Node.js code by doing:
```       
s3 =  AWSXRay.captureAWSClient(new S3({ apiVersion: '2006-03-01' }));
```
S3 now appears in the CloudWatch ServiceLens trace details.

The S3 bucket is configured to send S3 event notifications to SQS . According the X-Ray docs S3 passes down tracing headers to downstream services such as SQS. However this this appears not to be working as the SQS does not appear in the trace details, there is nothing beyond S3 in the tracing details.

**Comment by lupengamzn** [comment]
Hey @afayes ,

Based on the documentation, S3 will [propagate](https://docs.aws.amazon.com/xray/latest/devguide/xray-services-s3.html.) the X-Ray trace header through http header. Can you try parsing the `X-Amzn-Trace-Id` on SQS and manually [propagating](https://docs.aws.amazon.com/xray/latest/devguide/xray-services-sqs.html#xray-services-sqs-retrieving) it to downstream services to see if it works?

**Comment by afayes** [comment]
@lupengamzn I agree that based on the doc S3 will propagate the X-Ray trace header to SQS.

I cannot manually parse the X-Amzn-Trace-Id on SQS because the SQS is configured as an S3 event notification destination and there is no way to intercept the flow AFAIK unless I put a Lambda in between which is not an option in the project.

Could it be that the SQS queues were created before the functionality was available for S3 to SQS X-Ray integration?

Another question is where can I see the trace headers in the S3 object, is it supposed to be stored as message attributes which is not happenning?

**Comment by afayes** [comment]
Has there been any investigation on this?

**Comment by willarmiros** [comment]
Hi @afayes,

The reason you couldn't see the downstream SQS queue is because SQS and S3 are only integrated with X-Ray to propagate trace context through their systems, but neither system actually emits segment data. So you could see the S3 node because it's emitted from the upstream lambda function as a subsegment, but because neither the S3 nor SQS services could emit any data to represent the SQS queue, it could not show up on your service map.

**Comment by NathanielRN** [comment]
To add on to what @willarmiros said, X-Ray is able to infer the S3 node because Lambda (which is actively sending traces to X-Ray) knows that it called S3 as the next downstream service, but it does not know that S3 then calls SQS so it cannot infer that.

It's not until you're back at API Gateway that you are again actively sending traces to X-Ray and a segment is sent to X-Ray which can be connected to the previously inferred S3 node.

**Comment by willarmiros** [comment]
That is a fair callout @afayes, I've updated the documentation to be more clear. The context *is* propagated from S3 -> SQS to be read by a consumer on the other end, but it is disappointing that SQS cannot yet be seen on the trace.
